JoinColony/colonyCDapp
# Fix: Clickable transactions

## Description

- Refactor the way clickable transactions works, and ensure that it works for all transaction types.

An issue was raised (https://github.com/JoinColony/colonyCDapp/issues/3765) where motion related transactions were not opening the correct action when clicked. After investigating the issue, it turned out that most transactions were not opening the correct action, and in fact were not even opening the correct colony. This PR aims to refactor the way clicking on a transaction in the user hub works, and introduces a new 'associatedActionId' property on the transaction type. This will hold the action id (which is itself a transaction hash) of the action which should be opened. If it is not present, it can be assumed that the transaction hash contained in the transaction itself is the correct one.

This PR fixes the following navigation bugs when a transaction is clicked:
- Opening the correct colony for all transactions
- Going to the correct action for motion related transactions
- Going to the correct action for expenditure related transactions
- Going to the correct action for multi sig related transactions
- Going to the correct action for expenditure funding related transactions (permissions, multi sig, and repuation)
- Going to the correct colony dashboard for create colony transactions
- Going to the extensions page for extension related transactions
- Going to the incoming funds page for funds claimed transactions

## Testing

Theoretically, every transaction type can be tested, so if you really really want to spend two days testing this then feel free to just trigger every possible type of transaction method listed here: `src/types/transactions.ts` and make sure when clicking them in the user hub that you navigate where you would expect!

In reality of course, you probably don't want to do that, so here's some good places to test. Feel free to test either all of them or just some of them:

* After running the create data script and before doing anything, just check that all the transactions generated by that script running navigate correctly:

https://github.com/user-attachments/assets/35be96f7-4181-4df8-8643-55201470a9f8

* Install some extensions (I used reputation and multisig, and changed the multisig threshold since this will help with testing the motions further down the line). Now check the transactions for them:

https://github.com/user-attachments/assets/e2a81b37-1a9c-487c-832b-dbf191df42d3

* Create an agreement, and follow the entire flow until it has succeeded (stake for and against to trigger a vote, and then vote for it, reveal and finalize) and then check that these transactions navigate properly:

In my video some other motion related bugs are shown, but you can ignore these (claiming twice at the end, seeing a flash of the change vote button, and seeing oppose wins for a second even though the motion passed)

https://github.com/user-attachments/assets/05735ace-0d8f-42b7-a70e-9d7518d01f22

* Create a multisig motion (don't forget to first give permissions!), and again follow the flow until it has succeeded. In my testing, I just manipulated the multisig threshold instead of using two users, but feel free to test however you see fit. Check the transactions for these:

https://github.com/user-attachments/assets/6b5b14fc-6c77-44e9-845c-57327c42e669

* Create an expenditure (an advanced payment or a streaming payment, or a split payment) and again follow the flow until it has succeeded. It would be good to test funding with all different types of funding methods. In my video I've used reputation. Check the transactions for these:

https://github.com/user-attachments/assets/46b4d975-e20c-4710-b1d6-86ddaadd76b4

If you can think of any other transactions to test, feel free! Some other things which I haven't mentioned are cancelling transactions, and doing actions using permissions for example.

## Diffs

**New stuff** ✨

* New optional property `associatedActionId` on the transaction model

**Changes** 🏗

* Ensured that associatedActionId is added to transactions where necessary.
* Changed the way clickable transactions are handled.

Resolves #3765
